recipe-dsl/src/main/scala/com/ing/baker/recipe/common/InteractionFailureStrategy.scala

```diff
@@ -11,54 +11,93 @@ object InteractionFailureStrategy {
+                                private var maxTimeBetweenRetries: Option[Duration] = None,
+                                private var fireRetryExhaustedEvent: Option[String] = None) {
 
-      @tailrec def calculateMaxRetries(lastDelay: Duration,
-                                       backoffFactor: Double,
-                                       deadline: Duration,
-                                       totalDelay: Duration,
-                                       timesCounter: Int): Int = {
+      def withInitialDelay(initialDelay: Duration) = this.copy(initialDelay = initialDelay)
 
-        def min(d1Option: Option[Duration], d2: Duration): Duration = d1Option match {
-          case Some(d) if d < d2 => d
-          case _ => d2
-        }
+      def withBackoffFactor(backoffFactor: Double) = this.copy(backoffFactor = backoffFactor)
 
-        val nextDelay = min(maxTimeBetweenRetries, lastDelay * backoffFactor)
+      def withMaximumRetries(maximumRetries: Int) = this.copy(maximumRetries = Some(maximumRetries))
 
-        if ((totalDelay + nextDelay) > deadline) timesCounter
-        else calculateMaxRetries(nextDelay, backoffFactor, deadline, totalDelay + nextDelay, timesCounter + 1)
-      }
+      def withMaxTimeBetweenRetries(maxTimeBetweenRetries: Duration) = this.copy(maxTimeBetweenRetries = Some(maxTimeBetweenRetries))
+
+      def withFireRetryExhaustedEvent(fireRetryExhaustedEvent: String) = this.copy(fireRetryExhaustedEvent = Some(fireRetryExhaustedEvent))
+
+      def withFireRetryExhaustedEvent(fireRetryExhaustedEvent: Boolean) = this.copy(fireRetryExhaustedEvent = Some(defaultEventExhaustedName))
+
+      def withFireRetryExhaustedEvent(fireRetryExhaustedEvent: Event) = this.copy(fireRetryExhaustedEvent = Some(fireRetryExhaustedEvent.name))
+
+      def withDeadline(deadline: Duration) = this.copy(deadline = Some(deadline))
+
+      def build(): RetryWithIncrementalBackoff = {
+        if (deadline.isDefined) {
+          require(deadline.get > initialDelay, "deadline should be greater then initialDelay")
+
+          @tailrec def calculateMaxRetries(lastDelay: Duration,
+                                           backoffFactor: Double,
+                                           deadline: Duration,
+                                           totalDelay: Duration,
+                                           timesCounter: Int): Int = {
 
-      new RetryWithIncrementalBackoff(
-        initialTimeout = initialDelay,
-        backoffFactor,
-        maximumRetries = calculateMaxRetries(
-          lastDelay = initialDelay,
-          backoffFactor,
-          deadline,
-          totalDelay = initialDelay,
-          timesCounter = 1),
-        maxTimeBetweenRetries,
-        fireRetryExhaustedEvent)
+            def min(d1Option: Option[Duration], d2: Duration): Duration = d1Option match {
+              case Some(d) if d < d2 => d
+              case _ => d2
+            }
```

Comment:
```scala
def minOption(d1Option: Option[Duration], d2: Duration): Duration = d1Option.getOrElse(d2).min(d2)
```

Reply:
This is more readable but has 1 more cycle if the Option is None.
Then d2.min(d2) will be called.
still, lets change it since this is more readable.